release-25.4: sql: fix COPY with some concurrent schema changes

yuzefovich opened the pull request:
Backport 1/1 commits from #148549 on behalf of @yuzefovich.

----

Previously, if we had COPY running concurrently with some schema changes
that require usage of Puts (via ForcePut on the index), like ALTER
PRIMARY KEY, we could hit an internal error because the vectorized
encoder didn't support it. This commit adds the support for it, but it
also actually disables using the vectorized encoder for COPY when
ForcePut option is observed on at least one writable index. This is the
case since I observed many different corruption-like failures when
stressing the extended test, which aren't present if we use the
row-by-row insert.

Existing `TestLargeCopy` has been extended so that sometimes it also
performs ALTER PRIMARY KEY concurrently with the COPY as well as
randomized `copy_from_atomic_enabled` value. Note that in some cases
we're hitting "duplicate key" violation while there is a schema change,
and it seems somewhat expected to me, so I didn't look deeper and simply
made the test ignore such errors.

Fixes: #147955.

Release note (bug fix): CockroachDB could previously encounter "vector
encoder doesn't support ForcePut yet" error when performing COPY command
concurrently with some schema changes. The bug has been present since
before 23.2 and is now fixed.

----

Release justification: bug fix.

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @michae2 reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157200#-:-OdtFhvs7cVv5s-rue3J:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: fix COPY with some concurrent schema changes

Previously, if we had COPY running concurrently with some schema changes
that require usage of Puts (via ForcePut on the index), like ALTER
PRIMARY KEY, we could hit an internal error because the vectorized
encoder didn't support it. This commit adds the support for it, but it
also actually disables using the vectorized encoder for COPY when
ForcePut option is observed on at least one writable index. This is the
case since I observed many different corruption-like failures when
stressing the extended test, which aren't present if we use the
row-by-row insert.

Existing `TestLargeCopy` has been extended so that sometimes it also
performs ALTER PRIMARY KEY concurrently with the COPY as well as
randomized `copy_from_atomic_enabled` value. Note that in some cases
we're hitting "duplicate key" violation while there is a schema change,
and it seems somewhat expected to me, so I didn't look deeper and simply
made the test ignore such errors.

Release note (bug fix): CockroachDB could previously encounter "vector
encoder doesn't support ForcePut yet" error when performing COPY command
concurrently with some schema changes. The bug has been present since
before 23.2 and is now fixed.

yuzefovich commented:
I reverted the test code changes since there appear to be more failures (at least on 25.3 they were different that I hadn't seen previously).
